internal/pkg/cli/task_run_test.go

```diff
@@ -0,0 +1,316 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package cli
+
+import (
+	"errors"
+	"fmt"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/cli/mocks"
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/config"
+	"github.com/golang/mock/gomock"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+type basicOpts struct {
+	inNum    uint8
+	inCPU    uint16
+	inMemory uint16
+}
+
+func TestTaskRunOpts_Validate(t *testing.T) {
+	defaultOpts := basicOpts{
+		inNum:    1,
+		inCPU:    256,
+		inMemory: 512,
+	}
+
+	testCases := map[string]struct {
+		basicOpts
+
+		inImage          string
+		inDockerfilePath string
+
+		inTaskRole string
+
+		inApp              string
+		inEnv              string
+		inSubnetID         string
+		inSecurityGroupIDs []string
+
+		inEnvVars  map[string]string
+		inCommands string
+
+		appName string
+
+		mockStore      func(m *mocks.Mockstore)
+		mockFileSystem func(mockFS afero.Fs)
+
+		wantedError error
+	}{
+		"valid with no flag": {
+			basicOpts:   defaultOpts,
+			wantedError: nil,
+		},
+		"valid image and env": {
+			basicOpts: defaultOpts,
+
+			inImage:    "113459295.dkr.ecr.ap-northeast-1.amazonaws.com/my-app",
+			inTaskRole: "exec-role",
+
+			inEnv: "dev",
+
+			inEnvVars: map[string]string{
+				"NAME": "my-app",
+				"ENV":  "dev",
+			},
+			inCommands: "echo \"docker commands\"",
+
+			appName: "my-app",
+			mockStore: func(m *mocks.Mockstore) {
+				m.EXPECT().GetEnvironment("my-app", "dev").Return(&config.Environment{
+					App:  "my-app",
+					Name: "dev",
+				}, nil)
+			},
+
+			wantedError: nil,
+		},
+		"valid without image and env": {
+			basicOpts: defaultOpts,
+
+			inDockerfilePath: "hello/world/Dockerfile",
+			inTaskRole:       "exec-role",
+
+			inSubnetID:         "subnet-10d938jds",
+			inSecurityGroupIDs: []string{"sg-0d9sjdk", "sg-d33kds99"},
+
+			inEnvVars: map[string]string{
+				"NAME": "pj",
+				"ENV":  "dev",
+			},
+			inCommands: "echo \"docker commands\"",
+
+			mockFileSystem: func(mockFS afero.Fs) {
+				mockFS.MkdirAll("hello/world", 0755)
+				afero.WriteFile(mockFS, "hello/world/Dockerfile", []byte("FROM nginx"), 0644)
+			},
+			wantedError: nil,
+		},
+		"invalid number of tasks": {
+			basicOpts: basicOpts{
+				inNum:    0,
+				inCPU:    256,
+				inMemory: 512,
+			},
+			wantedError: errors.New("number of tasks must be positive"),
```

Comment:
Note: It is sometimes useful to extract error messages into a constant so they're reusable in the test -- that allows easier maintainability if we ever decide to change the error message.

Applies to other error test cases below.

Reply:
cool! I have extracted error messages for num, cpu and memory, etc. because their error message would appear more than once in the task_run file after I've implemented Ask